Feature read status for messages

juddruckers opened the pull request:
# What this does

1. Add new styles to login and sign up pages.
2. Consolidated Login and Signup components into one auth component because there were a lot of redundancies between the two.
3. Add `read` attribute to `Message` model
4. Add API end point and method on Message model to set read status to true
5. Add feature that notifies the current user if the other user in a conversation has read all messages sent by current user by displaying the other users avatar at the bottom of the active chat
6. Add feature that displays unread message count and bolds text if there are messages that are unread. Reset unread message count and unbold text to notify that user is up to date with that conversation.


# What it looks like 

### counter that displays how many unread messages in a conversation as well as bolded last message text.
![2021-11-27 12 20 46](https://user-images.githubusercontent.com/11431163/143724877-68e304d9-c0a9-4514-870f-11c35476bcd1.gif)

### reset unread messages when opening active chat and displaying if other user has read all messages 
![2021-11-27 12 21 39](https://user-images.githubusercontent.com/11431163/143724882-0a7e6d07-81b2-4445-a74e-b9e9abde9af8.gif)

### login and sign up design updates

![2021-11-27 12 27 02](https://user-images.githubusercontent.com/11431163/143724919-8c747381-adb2-4e87-abe3-9ae5eb9ae305.gif)

![2021-11-27 12 27 20](https://user-images.githubusercontent.com/11431163/143724932-a416b3a2-7842-45a4-a33d-2ce2196ee881.gif)

juddruckers committed:
feat: add prop-types dependency

juddruckers committed:
feat: add reusable text field component

juddruckers committed:
feat: Add styles for login and register pages.

Combined Login and Signup components into the Authentication component because there was a lot of redundancy between the two components.

Completes part 1 of 2 for issue #2

juddruckers committed:
feat: Add seen attribute to message model

juddruckers committed:
feat: add unread message count to conversations when retrieving users coversations list.

juddruckers committed:
feat: add badge to display amount of unread messages per conversation and bold preview text if conversation has unread messages

juddruckers committed:
feat: increment conversation.unreadMessages in store when receiving a new message to display unread message count on sidebar UI

juddruckers committed:
feat: Add messages endpoint and Message model method to update read status of messages.

juddruckers committed:
feat: Reset unread messages of conversation in store to 0 when opening a chat that has unread messages.

juddruckers committed:
feat: Add other user avatar to bottom of active chat messages when other user has read all messages sent by user.

Create resuable Avatar component used for OtherUserBubble component and to alert other user has read all messages sent by user

juddruckers committed:
refactor: Not necessary to check if every message has been read on Messages component. Only have to check if most recent message sent by user has been read by the other user in the conversation.

FilipeGorodscy reviewed (approved):
Great job!
- Awesome refactor to avoid duplication between Login/Signup pages, however, it is always best to stick with the issues mentioned in the ticket description! The bubble icon on the Sidebar is located under assets/images/bubble.svg
- Is your functionality working for increasing the unread count as soon as you create a new conversation?

Overall looks good! Feel free to merge the PR after addressing the comments!

juddruckers committed:
refactor:

refactor: Update messages according to their conversation ignored

Pass the conversation ID to the backend and update messages associated
to that conversation instead of calculating which messages should have
their read status updated to true on the client side.

Protect route from users updating messages who are not participants in a
conversation.

juddruckers committed:
refactor: Add white to theme palette and used theme.palette.white to replace css styles where white was declared as a string.

juddruckers committed:
refactor: Updated last read message feature to display the avatar under the last message that the other user has read so the current user knows where the other user stopped reading the conversation

juddruckers commented on client/src/components/ActiveChat/Messages.js:
```diff
@@ -1,20 +1,44 @@
-          <SenderBubble key={message.id} text={message.text} time={time} />
+          <>
+            <SenderBubble key={message.id} text={message.text} time={time} />
+            {index === messages.length - 1 && message.read && (
```
I'm sorry, I should have asked for more details on that feature. Looking back the image was vague. It has been fixed and the latest message read by the other user is determined on the backend.

juddruckers commented on client/src/components/ActiveChat/ActiveChat.js:
```diff
@@ -1,29 +1,50 @@
+        updateMessages(conversation.id, messagesToUpdate);
+      }
+    }
+  }, [activeConversation, conversation, updateMessages]);
```
Removed this functionality and updated it so that messages are updated based on the conversation ID

juddruckers committed:
refactor: small changes based suggestions from most recent PR review.
